File: BCW3/C++/Basic/Complex/Complex.cpp
#include <iostream>
#include <cmath>

#include "Complex.h"

Complex::Complex(double real, double imaginary) {
    this->real = real;
    this->imaginary = imaginary;
}

Complex::~Complex() {}
        
double Complex::getReal() const {
    return this->real;
}
double Complex::getImaginary() const {
    return this->imaginary;
}
        
bool Complex::operator==(const Complex& other) const {
    return other.real == this->real && other.imaginary == this->imaginary;
}
bool Complex::operator!=(const Complex& other) const {
    return !(*this == other);
}
void Complex::operator+=(const Complex& other) {
    this->real += other.real;
    this->imaginary += other.imaginary;
}
void Complex::operator-=(const Complex& other) {
    this->real -= other.real;
    this->imaginary -= other.imaginary;
}
Complex Complex::operator+(const Complex& other) const {
    Complex result = *this;
    result.real += other.real;
    result.imaginary += other.imaginary;
    return result;
}
Complex Complex::operator-(const Complex& other) const {
    Complex result = *this;
    result.real -= other.real;
    result.imaginary -= other.imaginary;
    return result;
}
Complex Complex::operator*(const Complex& other) const {
    Complex result;
    result.real = this->real * other.real - this->imaginary * other.imaginary;
    result.imaginary = this->real * other.imaginary + this->imaginary * other.real;
    return result;
}

std::ostream& operator<<(std::ostream& out, const Complex& complex) {
    if ( complex.getImaginary() < 0 ) {
        return std::cout << complex.getReal() << complex.getImaginary() << std::endl;
    } else {
        return std::cout << complex.getReal() << '+' << complex.getImaginary() << std::endl;
    }
}

File: BCW3/java/Basic/Complex/Complex.java
public class Complex implements Cloneable {
    private double real;
    private double imaginary;
    
    public Complex(double real, double imaginary) {
        this.real = real;
        this.imaginary = imaginary;
    }
    
    public Complex() {
        this(0.0, 0.0);
    }
    
    public double getReal() {
        return real;
    }
    
    public double getImaginary() {
        return imaginary;
    }
    
    public boolean equels(Complex other) {
        return real == other.real && imaginary == other.imaginary;
    }
    
    public Complex sum(Complex other) {
        return new Complex(real + other.real, imaginary + other.imaginary);
    }
    
    public Complex diff(Complex other) {
        return new Complex(real - other.real, imaginary - other.imaginary);
    }
    
    public Complex product(Complex other) {
        return new Complex(real * other.real - imaginary * other.imaginary, real * other.imaginary + imaginary * other.real);
    }
    
    public String toString() {
        if ( imaginary < 0 ) {
            return "(" + real + imaginary + "i)";
        } else {
            return "(" + real + "+" + imaginary + "i)";
        }
    }
    
    public Complex clone() throws CloneNotSupportedException {
        return (Complex) super.clone();
    }
}
